sbnci/Modules/RecoEff/RecoEff_module.cc

```diff
@@ -0,0 +1,417 @@
+      mc_energyEnd = particle->EndE();
+      mc_mass = particle->Mass();
+      mc_theta_xy = TMath::RadToDeg() * TMath::ATan(mc_pX0/mc_pY0);
+      mc_theta_yz = TMath::RadToDeg() * TMath::ATan(mc_pY0/mc_pZ0);
+      mc_theta_xz = TMath::RadToDeg() * TMath::ATan(mc_pX0/mc_pZ0);
+      mc_length = TrueTrackLength(particle);
+
+      if(n_tracks_map[mc_trackID] > 0 || n_showers_map[mc_trackID] > 0) reco_isReconstructed = true;
+      reco_nTracks = n_tracks_map[mc_trackID];
+      reco_nShowers = n_showers_map[mc_trackID];
+      reco_shower_purity = shower_pur_map[mc_trackID];
+      reco_shower_completeness = shower_comp_map[mc_trackID];
+      reco_track_purity = track_pur_map[mc_trackID];
+      reco_track_completeness = track_comp_map[mc_trackID];
+      reco_track_length = track_length_map[mc_trackID];
+      reco_shower_dEdx = shower_dEdx_map[mc_trackID];
+
+      fParticleTree->Fill();
+    }
+  }
+}
+
+std::vector<art::Ptr<recob::PFParticle> > RecoEff::GetPrimaryPFPs()
+{
+  std::vector<art::Ptr<recob::PFParticle> > primaries;
+  for(unsigned int pfp_i = 0; pfp_i < eHandlePFPs->size(); ++pfp_i) {
+    const art::Ptr<recob::PFParticle> pfp(eHandlePFPs,pfp_i);
+    if(pfp->IsPrimary()){
+      if(pfp->PdgCode() == 12 || pfp->PdgCode() == -12
+         || pfp->PdgCode() == 14 || pfp->PdgCode() == -14){
+        primaries.push_back(pfp);
+      }
+    }
+  }
+  return primaries;
+}
+
+art::Ptr<recob::PFParticle> RecoEff::GetPFP(long unsigned int const &id)
+{
+  const art::Ptr<recob::PFParticle> nullReturn;
+
+  for(unsigned int pfp_i = 0; pfp_i < eHandlePFPs->size(); ++pfp_i) {
+    const art::Ptr<recob::PFParticle> pfp(eHandlePFPs,pfp_i);
+    if(pfp->Self() == id) return pfp;
+  }
+  return nullReturn;
+}
+
+float RecoEff::Purity(std::vector< art::Ptr<recob::Hit> > const &objectHits, int const &trackID)
+{
+  std::map<int,int> objectHitsMap;
+
+  for(unsigned int i = 0; i < objectHits.size(); ++i) {
+    objectHitsMap[TruthMatchUtils::TrueParticleID(clockData,objectHits[i],true)]++;
+  }
+  return objectHitsMap[trackID]/static_cast<float>(objectHits.size());
+}
+
+float RecoEff::Completeness(std::vector< art::Ptr<recob::Hit> > const &objectHits, int const &trackID)
+{
+  std::map<int,int> objectHitsMap;
+
+  for(unsigned int i = 0; i < objectHits.size(); ++i) {
+    objectHitsMap[TruthMatchUtils::TrueParticleID(clockData,objectHits[i],true)]++;
+  }
+  return objectHitsMap[trackID]/static_cast<float>(hits_map[trackID]);
+}
+
+float RecoEff::TrackLength(art::Ptr<recob::Track> const &track)
+{
+  float length = 0;
+  int nTrajPoints = track->NumberTrajectoryPoints();
+
+  if(nTrajPoints < 2) return length;
+
+  for(int point = 1; point < nTrajPoints; ++point) {
+    TVector3 l = track->LocationAtPoint<TVector3>(point);
+    if(l.X() == -999 || l.Y() == -999 || l.Z() == -999) break;
+
+    TVector3 diff = track->LocationAtPoint<TVector3>(point) - track->LocationAtPoint<TVector3>(point-1);
+    length += TMath::Sqrt(diff.Mag2());
```

Comment:
I think TVector3 has a `Mag()` function.  Would that help here?

Reply:
Done [here](https://github.com/SBNSoftware/sbnci/pull/7/commits/e39fb816cd595c1be5af6af8f168c73d6c6e0fc1)

Reply:
following the link, the diff shows you removed the sqrt but forgot to remove the 2?

Reply:
Thanks Chris!